providers/adfs.go

```diff
@@ -0,0 +1,132 @@
+	*OIDCProvider
+}
+
+var _ Provider = (*OIDCProvider)(nil)
+
+const (
+	adfsProviderName = "Adfs"
+	adfsDefaultScope = "openid"
+)
+
+// NewADFSProvider initiates a new AdfsProvider
+func NewADFSProvider(p *OIDCProvider) *ADFSProvider {
+
+	p.ProviderData.setProviderDefaults(providerDefaults{
+		name:  adfsProviderName,
+		scope: adfsDefaultScope,
+	})
+	return &ADFSProvider{OIDCProvider: p}
+}
+
+// GetLoginURL overrides GetLoginURL to add Adfs parameters
+func (p *ADFSProvider) GetLoginURL(redirectURI, state string) string {
+	extraParams := url.Values{}
+	if p.ProtectedResource != nil && p.ProtectedResource.String() != "" {
+		extraParams.Add("resource", p.ProtectedResource.String())
+	}
+	a := makeLoginURL(p.ProviderData, redirectURI, state, extraParams)
+	return a.String()
+}
+
+// Redeem exchanges the OAuth2 authentication token for an Access\ID tokens
+func (p *ADFSProvider) Redeem(ctx context.Context, redirectURL, code string) (s *sessions.SessionState, err error) {
+	if code == "" {
+		return nil, ErrMissingCode
+	}
+	clientSecret, err := p.GetClientSecret()
+	if err != nil {
+		return nil, err
+	}
+
+	var jsonResponse struct {
+		AccessToken  string `json:"access_token"`
+		RefreshToken string `json:"refresh_token"`
+		ExpiresIn    int64  `json:"expires_in"`
+		IDToken      string `json:"id_token"`
+	}
+
+	params := url.Values{}
+	params.Add("client_id", p.ClientID)
+	params.Add("client_secret", clientSecret)
+	if p.ProtectedResource != nil && p.ProtectedResource.String() != "" {
+		params.Add("resource", p.ProtectedResource.String())
+	}
+	params.Add("grant_type", "authorization_code")
+	params.Add("code", code)
+	params.Add("redirect_uri", redirectURL)
+
+	err = requests.New(p.RedeemURL.String()).
+		WithContext(ctx).
+		WithMethod("POST").
+		WithBody(bytes.NewBufferString(params.Encode())).
+		SetHeader("Content-Type", "application/x-www-form-urlencoded").
+		Do().
+		UnmarshalInto(&jsonResponse)
+	if err != nil {
+		return
+	}
+
+	created := time.Now()
+	expires := time.Unix(jsonResponse.ExpiresIn, 0)
+
+	return &sessions.SessionState{
+		AccessToken:  jsonResponse.AccessToken,
+		IDToken:      jsonResponse.IDToken,
+		CreatedAt:    &created,
+		ExpiresOn:    &expires,
+		RefreshToken: jsonResponse.RefreshToken,
+	}, nil
+}
+
+// EnrichSession to add email
+func (p *ADFSProvider) EnrichSession(ctx context.Context, s *sessions.SessionState) error {
+	m, err := p.getEmailFromIDToken(ctx, s.IDToken)
+	if err != nil {
+		return err
+	}
+
+	s.Email = m
+
+	if s.Email == "" {
+		return errors.New("email not set")
+	}
+	return nil
+}
+
+func (p *ADFSProvider) getEmailFromIDToken(ctx context.Context, rawIDToken string) (string, error) {
+	idToken, err := p.Verifier.Verify(ctx, rawIDToken)
+	mail := ""
+	if err != nil {
+		return "", err
+	}
+
+	claims, err := p.getClaims(idToken)
+	if err != nil {
+		return "", fmt.Errorf("couldn't extract claims from id_token (%v)", err)
+	}
+
+	if m, ok := claims.raw["email"].(string); ok {
+		mail = m
```

Comment:
If you do `return m, nil` here, the logic in the following UPN alternative is an easier check (doesn't need the `mail == ""` part)

Reply:
I did it as you said in another commit, but then I got an error from codeclimate complaining about too much return statements